pkg/shared/families/malware/yara/yara.go

```diff
@@ -0,0 +1,168 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package yara
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+
+	"github.com/openclarity/kubeclarity/shared/pkg/job_manager"
+	"github.com/openclarity/kubeclarity/shared/pkg/utils"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/openclarity/vmclarity/pkg/shared/families/malware/common"
+	"github.com/openclarity/vmclarity/pkg/shared/families/malware/yara/config"
+	"github.com/openclarity/vmclarity/pkg/shared/families/malware/yara/util"
+	sharedutils "github.com/openclarity/vmclarity/pkg/shared/utils"
+)
+
+const (
+	ScannerName               = "yara"
+	TemporaryYaraRuleFilePath = "/tmp/yara-rules.yarac"
+)
+
+type Scanner struct {
+	name       string
+	logger     *log.Entry
+	config     config.Config
+	resultChan chan job_manager.Result
+}
+
+func (s *Scanner) Run(sourceType utils.SourceType, userInput string) error {
+	go func() {
+		retResults := common.Results{
+			Source:      userInput,
+			ScannerName: ScannerName,
+		}
+
+		if !s.isValidInputType(sourceType) {
+			retResults.Error = fmt.Errorf("received invalid input type for YARA scanner: %v", sourceType)
+			s.sendResults(retResults, nil)
+			return
+		}
+
+		s.logger.Debugf("Yara binary path: %s", s.config.YaraBinaryPath)
+		s.logger.Debugf("Yara rules URL: %s", s.config.YaraRuleURL)
+		rulesPath, err := s.getRuleFilePath()
+		if err != nil {
+			retResults.Error = fmt.Errorf("failed to get Yara rules file path: %v", err)
+		}
+		s.logger.Debugf("Yara rules file path: %s", rulesPath)
+
+		// Define the yara args to run
+		args := []string{"-C", rulesPath, "-r", "-w", "-m", userInput}
+
+		s.logger.Infof("Running yara...")
+		// Execute the yara command
+		// nolint:gosec
+		yaraCommand := exec.Command(s.config.YaraBinaryPath, args...)
+		out, err := sharedutils.RunCommand(yaraCommand)
+		if err != nil {
+			s.sendResults(retResults, fmt.Errorf("failed to run yara command: %s", err.Error()))
+		}
+
+		detectedMalware := util.ParseYaraScanOutput(string(out), s.logger)
+		retResults.Malware = detectedMalware
+		// In the case of Yara there is no scan summary.
+		retResults.Summary = &common.ScanSummary{}
+
+		s.sendResults(retResults, nil)
+	}()
+
+	return nil
+}
+
+func (s *Scanner) getRuleFilePath() (string, error) {
+	parsed, err := url.Parse(s.config.YaraRuleURL)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse Yara rule URL=%s: %v", s.config.YaraRuleURL, err)
+	}
+
+	switch parsed.Scheme {
+	case "file":
+		return parsed.Path, nil
+	case "http", "https":
+		return s.downloadRules()
+	default:
+		return "", errors.New("unsupported Yara rules URL")
+	}
+}
+
+func (s *Scanner) downloadRules() (string, error) {
+	s.logger.Infof("Downloading Yara rules...")
+	out, err := os.Create(TemporaryYaraRuleFilePath)
```

Comment:
what if already exist?

Reply:
We need to handle the scenario where the yara scanner called multiple-times (multiple inputs), but we probably only want to download the file once. Perhaps this should be downloaded as part of the scanners' initialisation instead of part of its "Run(..)" or wrapped in a "sync.Once(...)" so that it happens once per CLI run.

Reply:
@pbalogh-sa it doesn't look like the downloadRules function will handle multiple runs of the yara scanner, so if we have multiple inputs (multiple volumes for example) then it'll download the rules multiple times still. Are you going to address this in a follow up?

Reply:
On re-review I the called for downloadRules has been moved into the New function 👍